pkg/acpi/acpi.go

```diff
@@ -0,0 +1,219 @@
+)
+
+const (
+	// LengthOffset is the offset of the table length
+	LengthOffset = 4
+	// CSUMOffset is the offset of the single byte checksum in *most* ACPI tables
+	CSUMOffset = 9
+	// MinTableLength is the minimum length: 4 byte tag, 4 byte length, 1 byte revision, 1 byte checksum,
+	MinTableLength = 10
+)
+
+var (
+	// Debug implements fmt.Sprintf and can be used for debug printing
+	Debug        = func(string, ...interface{}) {}
+	unmarshalers = map[sig]func(Tabler) (Tabler, error){}
+)
+
+// addUnMarshaler is intended to be called by init functions
+// in this package. It adds an UnMarshaler for a given
+// ACPI signature.
+func addUnMarshaler(n string, f func(Tabler) (Tabler, error)) {
+	if _, ok := unmarshalers[sig(n)]; ok {
+		log.Fatalf("Can't add %s; already in use", n)
+	}
+	unmarshalers[sig(n)] = f
+}
+
+// GetHeader extracts a Header from a Tabler and returns a reference to it.
+func GetHeader(t Tabler) *Header {
+	return &Header{
+		Sig:             sig(t.Sig()),
+		Length:          t.Len(),
+		Revision:        t.Revision(),
+		CheckSum:        t.CheckSum(),
+		OEMID:           oem(t.OEMID()),
+		OEMTableID:      tableid(t.OEMTableID()),
+		OEMRevision:     t.OEMRevision(),
+		CreatorID:       t.CreatorID(),
+		CreatorRevision: t.CreatorRevision(),
+	}
+}
+
+// Flags takes 0 or more flags and produces a uint32 value.
+// Each argument represents one bit position, with the first flag
+// being bit 0. For each flag with a value of "1", the bit for that
+// flag will be set. For each flag with a value of "0", that bit in
+// the flag will be cleared.
+// This is mainly for consistency but it would allow
+// us in future to pass in an initial value and set or clear bits in it
+// depending on flags.
+// Current allowed values are "0" and "1". In future, if the flags are
+// not contiguous, we can allow "ignore" in future to ignore a flag.
+func flags(s ...flag) (uint8, error) {
+	var i, bit uint8
+	for _, f := range s {
+		switch f {
+		case "1":
+			i |= 1 << bit
+		case "0":
+			i &= ^(1 << bit)
+		default:
+			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valied", f)
+		}
+		bit++
+	}
+	return i, nil
+}
+
+// w writes 0 or more values to a bytes.Buffer, in LittleEndian order,
+func w(b *bytes.Buffer, val ...interface{}) {
+	for _, v := range val {
+		binary.Write(b, binary.LittleEndian, v)
+		Debug("\t %T %v b is %d bytes", v, v, b.Len())
+	}
+	Debug("w: done: b is %d bytes", b.Len())
+}
+
+// uw writes strings as unsigned words to a bytes.Buffer.
+// Currently it only supports 16, 32, and 64 bit writes.
+func uw(b *bytes.Buffer, s string, bits int) error {
+	// convenience case: if they don't set it, it comes in as "",
+	// take that to mean 0.
+	var v uint64
+	if s != "" {
+		var err error
+		if v, err = strconv.ParseUint(string(s), 0, bits); err != nil {
+			return err
+		}
+	}
+	switch bits {
+	case 8:
+		w(b, uint8(v))
+	case 16:
+		w(b, uint16(v))
+	case 32:
+		w(b, uint32(v))
+	case 64:
+		w(b, uint64(v))
```

Comment:
You can have a default here which returns an error or panics.

Reply:
done